[pcoq] Don't use physical equality to determine Pcoq.freeze worklist

ejgallego opened the pull request:
I've been observing huge parsing slowdowns in interactive use for a
while, and I finally found out that these are due to physical equality
in `share_tails` missing a lot of same-entries, thus creating a huge
work list for setting the parsing state in `Pcoq.freeze`.

This experimental commit uses a more precise equality, which is a 50x
win over a file interactive reparse, but indeed, it seems that we
should go for a more radical solution, which to really put the parsing
state in a reference we can just switch in O(1).

cc: @Alizter @herbelin maybe @ppedrot ?

ejgallego commented:
@coqbot bench now please

ppedrot commented on parsing/pcoq.ml:
```diff
@@ -432,10 +432,15 @@ type frozen_t =
+(* Physical equality here was missing a lot of same-entries, which
+   resulted in huge slowdown in Pcoq.freeze in some interactive users
+   *)
+let eq_grams x y = x = y
```
Using structural equality on arbitrary data is actually a regression. You can get unexpected results that break the abstraction barrier (e.g. if we store sets / maps in there) or outright anomalies.

Alizter replied:
It's not arbitrary however, isn't the type:
```
type frozen_t =
  (grammar_entry * GramState.t) list *
  CLexer.keyword_state
```

ppedrot replied:
Both GramState and grammar_entry can contain arbitary data.

Alizter replied:
What if we force them to provide their own hashing functions, that way we can hash and compare? This is admittedly quite some work, but something worth doing if we want to compare data structures seriously.

ppedrot replied:
Indeed, this is the correct solution. We need to store equality functions when defining new grammar extension states.

ejgallego replied:
Best would be to refactor the parse state so we don't need to do this huge hack. This has been one of the top priorities IMHO engineering side for a while.

ejgallego replied:
But yes, I wanted to try a quick PR to see how this would affect CI / bench, let me fix this part.

ejgallego replied:
You can't really see the slowdown unless you go to CoqIDE etc...

ejgallego replied:
> How does the physical equality get lost?

I don't know, that's a good question. Ping me on Zulip if you are curious about how to reproduce this (as it a bit involved)

ejgallego replied:
Uh uh, I talked too fast, this seems pretty hard to fix, how I'm supposed to compare `Dyn` and `Store.S.t` types ?

ejgallego replied:
The situation with this Dyn types is pretty precarious IMO, @ppedrot, already it is state that any state in the grammar thing must be marhsallable, so if so, it should be OK to use the polymorphic equality.

IMHO we should forbid Marshall and have types declare the proper interfaces.

ppedrot replied:
Just register a map associating an equality function to a given tag when registering new grammar state. I don't see anything precarious in there, and you don't have to modify Dyn in any way.

ejgallego replied:
I need to compare Dyn values too, maybe I'm missing something.

The type of thing to compare is `grammar_entry * GramState.t` , there are directly derived from `Store` and Dyn

ppedrot replied:
Comparing Dyn tags is easy, there is a function for that. Using it, and assuming you registered a function for each tag you can then check the contents of the values. For Store it's harder because the API doesn't expose an equality function and it hides the underlying Dyn code but it's actually easy to implement.

ppedrot replied:
... and rereading the code, you actually don't need to compare the GramState.t values. Indeed, the invariant is that this state can be obtained by repeatedly applying the actions from the first component starting from the empty state, so in some sense they are redundant. Thus, in the end, you just have to tweak `Pcoq.create_entry_command` and `Pcoq.create_grammar_command` to take an additional equality function (you can put it in a record as `entry_extension` and `grammar_extension` if you want) and use that to compare the values when sharing the tails.

ejgallego replied:
> you can put it in a record as entry_extension and grammar_extension if you want

I'll try that out thanks, but I'm wary about these records being marshalled.

ejgallego commented:
@SkySkimmer I tried your suggestion, and all the 4 possible combinations and unfortunately the equality is lost on both parts of the tuple :( This can add a large latency.

@ppedrot I've rebased and updated the comment, I'm unsure how to proceed with this unless you want me to significantly change Dyn and Store.

See my comment about the data already needing to be marshallable, so I guess poly eq is permissible here.

ppedrot commented:
Marshallable != equatable. Maps of marshallable data are marshallable, but structural equality is broken. Same thing for references, two references to the same value are at the same time marshallable and structurally equal.

ejgallego commented:
> Marshallable != equatable. Maps of marshallable data are marshallable, but structural equality is broken. Same thing for references, two references to the same value are at the same time marshallable and structurally equal.

Indeed, but spurious mis-equalities are not a risk of soundness here, as this is just an optimization not to redo all the state setting.

But I fully agree that this is not good, it is just that I'm a lost on how to fix this in other way. What do you suggest ?

ejgallego commented:
Concretely, we'd need to add equality for:
- `GrammarCommand.t` where `module GrammarCommand = Dyn.Make ()`
- `GramState.t` where `module GramState = Store.Make ()`

ppedrot commented:
@ejgallego see [ppedrot/coq#pcoq-factor-better-safe](https://github.com/ppedrot/coq/tree/pcoq-factor-better-safe). You just have to write the proper equality functions at the various FIXME, although it will probably work already better than currently because it only performs physical equality on the relevant subcomponents which are (maybe?) not changed across calls.

ppedrot commented:
Closing this PR, which is superseded by #16544.

ejgallego commented:
This PR was fixing an efficiency problem which #16544 doesn't yet, so the PR is not superseeded.

ppedrot commented:
This code is unsound as written currently, and in need of a rebase anyways. I think it would be better to start afresh, but your call.

ejgallego commented:
> This code is unsound as written currently, and in need of a rebase anyways. I think it would be better to start afresh, but your call.

Yes, it needs a rebase. Unsound it was not tho, where is the unsoundness (I don't consider an optimization potentially not triggering to be an unsoundness)

ejgallego commented:
We don't need to keep discussing this, but I'm unable to understand your example. If two arrays are equal structurally I would think that I can substitute one for the other unless some clients are doing really heathen things with physical equality (and IMHO that would be a greater problem), but let's move on tho.

ejgallego commented:
Actually the fix proposed in the other PR works, sorry for the noise and thanks!
